internal/check/engine.go

```diff
@@ -137,7 +137,12 @@ func (e *Engine) checkExpandSubject(r *relationTuple, restDepth int) checkgroup.
 
 		// If not, we must go another hop:
 		for _, result := range results {
-			innerCtx, visited = graph.CheckAndAddVisited(innerCtx, result.To.Subject)
+			sub := &relationtuple.SubjectSet{
+				Namespace: result.To.Namespace,
+				Object:    result.To.Object,
+				Relation:  result.To.Relation,
+			}
+			innerCtx, visited = graph.CheckAndAddVisited(innerCtx, sub)
```

Comment:
Good catch, at this point `result.To.Subject` is the same for all results: https://github.com/ory/keto/blob/db59a33ed171fbb56ed81c31e822199e1011910f/internal/persistence/sql/traverser.go#L108